### PR TITLE
CONV backward

### PR DESCRIPTION
**Pull request by weiyanlin117**
test passed 
only support weights gradient and output gradient, on bias gradient yet, 
Since I notice pytorch doesn't implement bias gradient, I treat bias gradient as optional? 
I can spend time if it is needed.

 
Screenshot 2023-07-26 at 11 54 48 PM

### CODE REVIEW

File: lib/nnc/cmd/convolution/mps/ccv_nnc_conv_mps.m [hunk 1]
```diff
@@ -110,3 +197,11 @@ static int _ccv_nnc_conv_forw(const ccv_nnc_cmd_t cmd, const ccv_nnc_hint_t hint
 	registry->exec = _ccv_nnc_conv_forw;
 }
 
+REGISTER_COMMAND_BACKEND(CCV_NNC_CONVOLUTION_BACKWARD, CCV_NNC_BACKEND_MPS)(ccv_nnc_cmd_backend_registry_t* const registry)
+{
+	registry->tensor_formats = CCV_TENSOR_FORMAT_NCHW;
```
Comment (liuliu) [inline lib/nnc/cmd/convolution/mps/ccv_nnc_conv_mps.m]: You might want to also support NHWC layout if MPSGraph side supports that, it is visibly faster.

**Review by liuliu — COMMENTED** [review]
bias term needs to be supported though I think.

**Comment by liuliu** [comment]
Have you tried NHWC layout (with weights potentially in NCHW layout)?

File: test/int/nnc/mpsdnn.tests.c [hunk 2]
```diff
@@ -1444,4 +1444,167 @@ TEST_CASE("mps leaky relu gradient in float")
...
+	for (i = 0; i < OUTPUT_SIZE * OUTPUT_SIZE * OUTPUT_DIM * ccv_max(1, BATCH_SIZE); i++)
+		g->data.f32[i] = dsfmt_genrand_open_close(&dsfmt) / OUTPUT_DIM; // (OUTPUT_SIZE * OUTPUT_SIZE * OUTPUT_DIM);
+	// Copy generated matrix values over to GPU.
+	ccv_nnc_tensor_t* ga = ccv_nnc_tensor_new(0, GPU_TENSOR_NHWC(000, 32F, BATCH_SIZE, INPUT_SIZE, INPUT_SIZE, INPUT_DIM), 0);
+	ccv_nnc_tensor_t* gg = ccv_nnc_tensor_new(0, GPU_TENSOR_NHWC(000, 32F, BATCH_SIZE, OUTPUT_SIZE, OUTPUT_SIZE, OUTPUT_DIM), 0);
+	ccv_nnc_tensor_t* gh = ccv_nnc_tensor_new(0, GPU_TENSOR_NHWC(000, 32F, BATCH_SIZE, INPUT_SIZE, INPUT_SIZE, INPUT_DIM), 0);
+	ccv_nnc_tensor_t* gw = ccv_nnc_tensor_new(0, GPU_TENSOR_NHWC(000, 32F, OUTPUT_DIM, KERNEL_SIZE, KERNEL_SIZE, INPUT_DIM), 0);
+	ccv_nnc_tensor_t* gbias = ccv_nnc_tensor_new(0, GPU_TENSOR_NHWC(000, 32F, 1, 1, 1, OUTPUT_DIM), 0);
+	ccv_nnc_tensor_t* gdw = ccv_nnc_tensor_new(0, GPU_TENSOR_NHWC(000, 32F, OUTPUT_DIM, KERNEL_SIZE, KERNEL_SIZE, INPUT_DIM), 0);
+	ccv_nnc_tensor_t* gdbias = ccv_nnc_tensor_new(0, GPU_TENSOR_NHWC(000, 32F, 1, 1, 1, OUTPUT_DIM), 0);
```
Comment (weiyanlin117) [inline test/int/nnc/mpsdnn.tests.c:1623]: for NHWC, I need to put output bias dim at the Channel explicitly
as  (1, 1, 1, OUTPUT_DIM)
with g (BATCH_SIZE, OUTPUT_SIZE, OUTPUT_SIZE, OUTPUT_DIM)
thus we will reduce to the OUTPUT_DIM.

**Comment by weiyanlin117** [comment]
Screenshot 2023-07-27 at 5 23 18 PM

**Comment by weiyanlin117** [comment]
NCHW works as same as previous
Screenshot 2023-07-27 at 5 24 19 PM

**Comment by liuliu** [comment]
Probably need a regen.

**Comment by weiyanlin117** [comment]
Screenshot 2023-07-28 at 11 27 57 AM

**Review by liuliu — COMMENTED** [review]
You missed gating on dw is not provided, h is not provided, or dbias is not provided. See the cpu_ref.c that gated these cases.

**Comment by weiyanlin117** [comment]
normal test passed.
commented out dbias test passed.

File: test/int/nnc/mpsdnn.tests.c (hunk 2 above)
Comment (liuliu) [reply test/int/nnc/mpsdnn.tests.c:1623]: You can do that within the MPS graph ... But I will merge this PR now and do that later.

**Review by liuliu — APPROVED** [review]

**Pull request merged**